Clarify that the range operator may allocate a new collection

BillWagner opened the pull request:
Fixes #17912
Fixex #20492

For may collections, the range operator allocates storage for the returned range. This has performance implications.

As part of this, remove the terms "subrange" and :slice" which could imply that the return does not allocate new storage and copy the range.

BillWagner committed:
Clarify that the range operator may allocate a new collection

Fixes #17912

For may collections, the range operator allocates storage for the returned range. This has performance implications.

As part of this, remove the terms "subrange" and :slice" which could imply that the return does not allocate new storage and copy the range.

svick commented on docs/csharp/tutorials/ranges-indexes.md:
```diff
@@ -72,10 +72,13 @@ The following sample shows many of the reasons for those choices. Modify `x`, `y
 Any type that provides an [indexer](../programming-guide/indexers/index.md) with an <xref:System.Index> or <xref:System.Range> parameter explicitly supports indices or ranges respectively. An indexer that takes a single <xref:System.Range> parameter may return a different sequence type, such as <xref:System.Span%601?displayProperty=nameWithType>.
 
+> [!IMPORTANT]
+> The range operator often allocates a copy to store the range of elements returned. This causes extra allocations and copies, impacting performance.
```
This makes it sound like you should avoid ranges altogether in performance sensitive code even though using them correctly with `Span` or `Memory` can improve performance. Maybe it could say something like this?

> Depending on the sequence type, the range operator can allocate a copy to store the range of elements returned. This causes extra allocations and copies, impacting performance. In performance sensitive code, consider using `Span<T>` or `Memory<T>` as the sequence type, since the range operator does not allocate for them.

pkulikov commented:
@BillWagner is this a good PR and article to cover #20492 as well?

BillWagner committed:
updated based on feedback.

Thanks @svick and @pkulikov These are excellent suggestions.

BillWagner commented:
closing and reopening for a fresh build.

svick commented on docs/csharp/tutorials/ranges-indexes.md:
```diff
@@ -77,7 +77,11 @@ Indexes and ranges provide clear, concise syntax to access a single element or a
 
 > [!IMPORTANT]
-> The range operator often allocates a copy to store the range of elements returned. This causes extra allocations and copies, impacting performance.
+> The performance of code using the range operator depends on the source collection operand.
+>
+> The time complexity of the range operator depends on the storage type. For example, if the storage is a `string` or an array, then the result is a copy of the specified section of the input, so the time complexity is *O(N)* (where N is the length of the range). On the other hand, if it's a <xref:System.Memory%601.Span%2A?displayProperty=nameWithType> or a <xref:System.Memory%601?displayProperty ...
+>
+> In addition to the time complexity, this causes extra allocations and copies, impacting performance. In performance sensitive code, consider using `Span<T>` or `Memory<T>` as the sequence type, since the range operator does not allocate for them.
```
This now talks at different points about "source collection", "storage type" and "sequence type" and I'm not sure it will be clear to the reader what the relationship between them is.

Can this be made mode consistent? The surrounding text mostly uses just "sequence", sometimes as "source sequence".

svick commented on docs/csharp/tutorials/ranges-indexes.md:
```diff
@@ -77,7 +77,11 @@ Indexes and ranges provide clear, concise syntax to access a single element or a
-> The range operator often allocates a copy to store the range of elements returned. This causes extra allocations and copies, impacting performance.
+> The performance of code using the range operator depends on the source collection operand.
+>
+> The time complexity of the range operator depends on the storage type. For example, if the storage is a `string` or an array, then the result is a copy of the specified section of the input, so the time complexity is *O(N)* (where N is the length of the range). On the other hand, if it's a <xref:System.Memory%601.Span%2A?displayProperty=nameWithType> or a <xref:System.Memory%601?displayProperty ...
```
I don't think this should reference the `Memory<T>.Span` property, but the `Span<T>` type.

```suggestion
> The time complexity of the range operator depends on the storage type. For example, if the storage is a `string` or an array, then the result is a copy of the specified section of the input, so the time complexity is *O(N)* (where N is the length of the range). On the other hand, if it's a <xref:System.Span%601?displayProperty=nameWithType> or a <xref:System.Memory%601?displayProperty=nameWithType>, the result references the same backing store, which means there is no copy and the operation is *O(1)*.
```

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Petr Onderka <gsvick@gmail.com>

BillWagner committed:
use "sequence" consistently

Feedback: "sequence", "source collection", and "storage type" were used for the sequence operand. Pick one.

svick commented on docs/csharp/tutorials/ranges-indexes.md:
```diff
@@ -77,11 +77,11 @@ Indexes and ranges provide clear, concise syntax to access a single element or a
+> The time complexity of the range operator depends on the sequence type. For example, if the sequence is a `string` or an array, then the result is a copy of the specified section of the input, so the time complexity is *O(N)* (where N is the length of the range). On the other hand, if it's a <xref:System.Span%601?displayProperty=nameWithType> or a <xref:System.Memory%601?displayProperty=nameWit ...
 >
-> In addition to the time complexity, this causes extra allocations and copies, impacting performance. In performance sensitive code, consider using `Span<T>` or `Memory<T>` as the sequence type, since the range operator does not allocate for them.
+> In addition to the ftime complexity, this causes extra allocations and copies, impacting performance. In performance sensitive code, consider using `Span<T>` or `Memory<T>` as the sequence type, since the range operator does not allocate for them.
```
```suggestion
> In addition to the time complexity, this causes extra allocations and copies, impacting performance. In performance sensitive code, consider using `Span<T>` or `Memory<T>` as the sequence type, since the range operator does not allocate for them.
```

svick reviewed (approved):
Just a typo, otherwise LGTM.

BillWagner committed:
Update docs/csharp/tutorials/ranges-indexes.md

Co-authored-by: Petr Onderka <gsvick@gmail.com>

luisquintanilla reviewed (approved):
Looks good to me. I learned something new today 😃
